File: backend/routes/navigation/alg.js
const { PrismaClient } = require('../../prisma/generated/prisma');
const prisma = new PrismaClient()

// Utility to build the adjacency list (graph) from routes
async function buildGraph() {
    const routes = await prisma.routes.findMany({
        select: { departure: true, destination: true }
    });
    const graph = {};
    for (const route of routes) {
        if (!graph[route.departure]) graph[route.departure] = [];
        graph[route.departure].push(route.destination);
    }
    return graph;
}

// BFS to find a path from start to end
async function findRoutePath(startId, endId) {
    const graph = await buildGraph();
    const queue = [[startId]];
    const visited = new Set();

    while (queue.length > 0) {
        const path = queue.shift();
        const node = path[path.length - 1];
        if (node === endId) return path; // Found a path!
        if (visited.has(node)) continue;
        visited.add(node);
        const neighbors = graph[node] || [];
        for (const neighbor of neighbors) {
            queue.push([...path, neighbor]);
        }
    }
    return null; // No path found
}

// Breadth-First Search (BFS)
async function findRoute(from, to) {
    try {
        const path = await findRoutePath(parseInt(from, 10), parseInt(to, 10));
        if (!path) {
            return 0;
        }

        // Optionally, get airport details for each stop
        const airports = await prisma.airports.findMany({
            where: { id: { in: path } },
            select: { id: true, name: true, city: true, country: true, time_zone: true }
        });
        // Sort airports in path order
        const airportMap = Object.fromEntries(airports.map(a => [a.id, a]));
        const orderedAirports = path.map(id => airportMap[id]);

        return orderedAirports;
    } catch (error) {
        return error;
    }
};

// Exports the function to be used in the route handler
module.exports = findRoute;

File: backend/routes/navigation/alg.py
from ORM.db import get_db
from ORM.models import Routes, Airports
from collections import deque

# Utility to build the adjacency list (graph) from routes
def buildGraph():
    with get_db() as db:
        routes = db.query(Routes.departure, Routes.destination).all()
        graph = {}
        for departure, destination in routes:
            if departure not in graph:
                graph[departure] = []
            graph[departure].append(destination)
        return graph

# BFS to find a path from start to end
def findRoutePath(startId, endId):
    graph = buildGraph()
    queue = deque([[startId]])
    visited = set()

    while queue:
        path = queue.popleft()
        node = path[-1]
        if node == endId:
            return path  # Found a path!
        if node in visited:
            continue
        visited.add(node)
        neighbors = graph.get(node, [])
        for neighbor in neighbors:
            queue.append(path + [neighbor])
    return None  # No path found

# Breadth-First Search (BFS)
def findRoute(fromId, toId):
    try:
        path = findRoutePath(int(fromId), int(toId))
        print("Found path:\n")
        print(path)
        if not path:
            return 0

        with get_db() as db:
            airports = db.query(Airports).filter(Airports.id.in_(path)).all()
            airport_map = {a.id: a for a in airports}
            ordered_airports = [airport_map[id] for id in path]

            return ordered_airports
    except Exception as error:
        return 0
